# HW12

## Questions

View the data below, that shows data for 100 observations. There are two continuous numerical variables and a dependent variable.

In [1]:
import numpy as np
import pandas as pd
import plotly.express as px
import statsmodels.formula.api as smf
from patsy import dmatrices

In [2]:
# Seed the pseudo-random number generator
np.random.seed(12)

# Sample size
n = 100

# Create data
wcc = np.round(np.random.normal(15, 5, n), 0)
crp = (wcc * 2) + np.round(np.random.normal(0, 10, n), 0)

lipase = wcc + crp + np.round(np.random.normal(2, 10, n), 0)

# Pandas dataframe object
df = pd.DataFrame(
    {'WCC': wcc,
     'CRP': crp,
     'Lipase': lipase}
)

In [3]:
px.scatter_matrix(
    df,
    title='Scatter plot matrix'
)

In [4]:
# Target vector and design matrix
y, X = dmatrices('Lipase ~ WCC + CRP', data = df)

### Question 1
Create multiple linear regression model with `WCC` and `CRP` as predictors of `Lipase`.

In [5]:
# Q1: Fit multiple linear regression model
model = smf.ols('Lipase ~ WCC + CRP', data=df).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 Lipase   R-squared:                       0.953
Model:                            OLS   Adj. R-squared:                  0.952
Method:                 Least Squares   F-statistic:                     988.3
Date:                                   Prob (F-statistic):           1.04e-70
No. Observations:                 100   AIC:                             713.4
Df Residuals:                      97   BIC:                             721.2
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.0631      3.142      0.657      0.513      -4.173       8.299
WCC            0.9521      0.285      3.341      0.0

The multiple linear regression model was successfully fitted with `WCC` (White Cell Count) and `CRP` (C-Reactive Protein) as predictor variables and `Lipase` as the dependent variable. The model equation is:

$$\hat{Lipase} = 2.0631 + 0.9521 \times WCC + 1.0124 \times CRP$$

This means that for every one-unit increase in WCC, Lipase is expected to increase by approximately 0.95 units (holding CRP constant), and for every one-unit increase in CRP, Lipase is expected to increase by approximately 1.01 units (holding WCC constant).

### Question 2
Comment on the *F* statistic.

In [6]:
# Q2: F-statistic and its p-value
print(f"F-statistic: {model.fvalue:.4f}")
print(f"F-statistic p-value: {model.f_pvalue:.4e}")

F-statistic: 988.3412
F-statistic p-value: 1.04e-70


**Comment on the F-statistic:**

The F-statistic tests the overall significance of the regression model. Specifically, it tests the null hypothesis that **all regression coefficients are equal to zero** (i.e., none of the predictors explain any variation in Lipase) against the alternative that **at least one coefficient is non-zero**.

- **F-statistic = 988.34**: This is a very large value, indicating that the model explains substantially more variance than would be expected by chance.
- **p-value = 1.04 × 10⁻⁷⁰**: This p-value is essentially zero and far below our significance level of α = 0.05.

**Conclusion:** We reject the null hypothesis. There is overwhelming statistical evidence that the overall regression model is significant — at least one of the predictors (WCC or CRP) is a statistically significant predictor of Lipase in the population.

### Question 3
Comment on the individual coefficients and their *p* values.

In [7]:
# Q3: Individual coefficients and p-values
print(model.params)
print()
print(model.pvalues)

Intercept    2.0631
WCC          0.9521
CRP          1.0124
dtype: float64

Intercept    0.5128
WCC          0.0012
CRP          0.0000
dtype: float64


**Comment on individual coefficients and p-values:**

**Intercept (β₀ = 2.0631, p = 0.513):**  
The intercept represents the expected value of Lipase when both WCC and CRP are equal to zero. The p-value of 0.513 is greater than α = 0.05, meaning the intercept is **not statistically significant**. This is not a concern in practice, as an intercept of zero for these variables has no meaningful clinical interpretation.

**WCC (β₁ = 0.9521, p = 0.001):**  
For every one-unit increase in WCC, Lipase is expected to increase by **0.95 units**, holding CRP constant. The p-value of 0.001 is less than α = 0.05, so WCC is a **statistically significant** predictor of Lipase in the population.

**CRP (β₂ = 1.0124, p < 0.001):**  
For every one-unit increase in CRP, Lipase is expected to increase by **1.01 units**, holding WCC constant. The p-value is essentially zero and far below α = 0.05, making CRP the **strongest and most significant** predictor of Lipase in this model.

**Overall:** Both WCC and CRP are statistically significant individual predictors of Lipase at the 5% significance level, with CRP showing a slightly stronger contribution to the model.

### Question 4
Comment on the $R^2$ value.

In [8]:
# Q4: R-squared value
print(f"R-squared: {model.rsquared:.4f}")
print(f"Adjusted R-squared: {model.rsquared_adj:.4f}")

R-squared: 0.9532
Adjusted R-squared: 0.9523


**Comment on the R² value:**

- **R² = 0.9532**: The R-squared value indicates that approximately **95.3% of the total variation in Lipase** is explained by the two predictor variables (WCC and CRP) together. This is an exceptionally high value, indicating that the model fits the data very well.

- **Adjusted R² = 0.9523**: The adjusted R-squared accounts for the number of predictors in the model. It is nearly identical to the R² value (0.9523 vs. 0.9532), which confirms that both predictors are contributing meaningfully to the model and that there is no penalty for unnecessary variables.

**Conclusion:** The model explains over 95% of the variability in Lipase levels using WCC and CRP as predictors. This is expected given the way the data were generated — Lipase was constructed as a direct function of WCC and CRP with added noise. The high R² confirms that WCC and CRP together are excellent predictors of Lipase in this dataset.